### DISCUSSION TITLE
SSL Issue when trying to append values using FreeRTOS.

### DISCUSSION THREAD

**Discussion by roshanyadav29** [mobizt/ESP-Google-Sheet-Client#40] [Q&A] [answered]
I am getting this error when trying to append the values. Can anyone help me?
17:28:11.740 -> > ERROR.mRunUntil: SSL internals timed out!
17:28:11.740 -> > ERROR.mConnectSSL: Failed to initlalize the SSL layer.
17:28:11.740 -> > ERROR.mConnectSSL: Unknown error code.

Is there a way to manually close or restart the connection within the code?

I am using ESP32C3 Board with Arduino IDE v2.3.2. I am trying to implement FreeRTOS-based tasks to connect to Wi-Fi, get the time and data, process it and then send it to Google Sheets. Any Leads to why I am getting this error would be greatly appreciated

**Comment by mobizt** [comment]
If you set the server or root CA certificate, that certificate is expired. You have to provide a new certificate.

If you don't know how, you should not use `GSheet.setCert`.

**Reply by roshanyadav29** [reply]
Thanks for the reply. I don't know about server or root CA Certificates. I am also not using GSheet.setCert as far as I can tell.

```

void setup_sheets() {

  // Set the callback for Google API access token generation status (for debug only)
  GSheet.setTokenCallback(tokenStatusCallback);

  // Set the seconds to refresh the auth token before expire (60 to 3540, default is 300 seconds)
  GSheet.setPrerefreshSeconds(10 * 60);

  // Begin the access token generation for Google API authentication
  GSheet.begin(CLIENT_EMAIL, PROJECT_ID, PRIVATE_KEY);

  GSheet.setSystemTime(getTime());

  if (!check_sheets_ready())  // Return 0 if not ready
  {
#if __TEMP_DEBUG2__
    Serial.println(FILE_NAME_GSHEET "Not available");
#endif
  } else {
#if __TEMP_DEBUG2__
    Serial.println(FILE_NAME_GSHEET "Available");
#endif
  }
}

bool check_sheets_ready() {
  bool ready = GSheet.ready();
  return ready;
}
```

This is how I am initializing the Google Sheets. then after that I am just using append function to write values to the sheets.

```
void append_values_apm10(String __date, String __time, float __Value_PM1, float __Value_PM2_5, float __Value_PM10, int __AQI_EPA_2_5, int __AQI_EPA_10, int __AQI_CPCB_2_5, int __AQI_CPCB_10) {

  FirebaseJson response;

  Serial.println("\nAppend spreadsheet values...");
  Serial.println("----------------------------");

  FirebaseJson valueRange;

  valueRange.add("majorDimension", "COLUMNS");
  valueRange.set("values/[0]/[0]", __date);
  valueRange.set("values/[1]/[0]", __time);
  valueRange.set("values/[2]/[0]", __Value_PM1);
  valueRange.set("values/[3]/[0]", __Value_PM2_5);
  valueRange.set("values/[4]/[0]", __Value_PM10);
  valueRange.set("values/[5]/[0]", __AQI_EPA_2_5);
  valueRange.set("values/[6]/[0]", __AQI_EPA_10);
  valueRange.set("values/[7]/[0]", __AQI_CPCB_2_5);
  valueRange.set("values/[8]/[0]", __AQI_CPCB_10);

  // For Google Sheet API ref doc, go to https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets.values/append
  // Append values to the spreadsheet
  bool success = GSheet.values.append(&response /* returned response */, spreadsheetId /* spreadsheet Id to append */, "APM10!A1" /* range to append */, &valueRange /* data range to append */);
  if (success) {
    response.toString(Serial, true);
    valueRange.clear();
  } else {
    Serial.println(GSheet.errorReason());
  }
  Serial.println();
  Serial.println(ESP.getFreeHeap());
}
```

This is the function to append the data that I am using. Took it from an example sketch. As for calling this particular function, I have created a rtos based timer that when reaches zero would call the function to append the values. Did I miss anything in the implementation? When using Arduino IDE Super Loop approach it's working great. Having trouble with the FreeRTOS method.

**Reply by roshanyadav29** [reply]
If it helps, the error is followed by this:

```
19:22:29.656 -> FROM PROCESS : Read Sensor data, state -> 1
19:22:29.656 -> FROM PROCESS : Calculate AQI from Sensor data, state -> 2
19:22:29.656 -> FROM PROCESS : Publish Sensor data, state -> 3
19:22:29.656 -> 
19:22:29.656 -> Append spreadsheet values...
19:22:29.656 -> ----------------------------
19:23:30.192 -> > ERROR.mRunUntil: SSL internals timed out!
19:23:30.192 -> > ERROR.mConnectSSL: Failed to initlalize the SSL layer.
19:23:30.192 -> > ERROR.mConnectSSL: Unknown error code.
19:23:30.238 -> unknown error
19:23:30.238 -> 
19:23:30.238 -> 141364
19:23:30.238 -> 
19:23:30.238 -> Append spreadsheet values...
19:23:30.238 -> ----------------------------
19:23:30.280 -> send request failed
19:23:30.280 -> 
19:23:30.280 -> 143704
19:23:30.280 -> 
19:23:30.280 -> Append spreadsheet values...
19:23:30.280 -> ----------------------------
```

**Reply by mobizt** [reply]
You must run the networking task only on CPU core 1 and should use `vTaskDelay` to allow system task to run.

**Reply by roshanyadav29** [reply]
Thanks again for replying. I am using only the default core for all tasks since the esp32c3 chip only has a single core. Regarding the delay, since that function is only called once every 10 seconds using the timer I thought I wouldn't need it. I will trying including and playing with the delay.

**Reply by mobizt** [reply]
It depends on how you run the task. The delay was needed for `FreeRTOS` tasks if your code runs in the infinite loop in the `FreeRTOS` task to allow system tasks to run included wdt feeding. 

Why have you run your code in FreeRTOS task in single core MCU?

**Reply by mobizt** [reply]
You have to be familiar with the ESP32 platform you used.

In this case is user code issue not a library issue.

**Reply by roshanyadav29** [reply]
A fair question, the simple answer would be to just test and learn more about FreeRTOS. Idea is, I have to acquire samples from multiple sensors so I thought I could separate them as multiple tasks and then maybe make a few calculations like averaging or something, and then when the RTOS timer is up I will call the append function to push the data to google sheets.

When is use the loop approach there is a bit of delay associated that gets added when posting this data. So I thought I could use the tasks approach (FreeRTOS) and maybe eliminate that as well as learn something new.

**Reply by roshanyadav29** [reply]
> You have to be familiar with the ESP32 platform you used.
> 
> In this case is user code issue not a library issue.

Sorry didn't mean it like that. I thought the error would help me identify the issue of why SSL is timing out, hence the reason why I post a query here.

**Reply by mobizt** [reply]
You can test the networking in your FreeRTOS task with basic https server connection using WiFiClientSecure to check whether it works or not inside the FreeRTOS task.

For this library, you should create minimal but complete code that can reproduce the issue and post here that I can check and test.

I have ESP32C3 but I'm not on my device and  PC now.

**Reply by roshanyadav29** [reply]
Understood. I will try it out with WiFiClientSecure as you mentioned. I will update after testing, regarding the basic https server connection using the example codes. Thank you for your suggestions as well as quick replies.

**Reply by mobizt** [reply]
I read your post again.

If you are using a timer, you should avoid running blocking and memory consuming codes inside its ISR or callback function.

I do not recommend using this library or any networking library inside the ISR.

You should copy or save the timer status in global variables to check and run the library in the main loop instead.

**Reply by roshanyadav29** [reply]
Sorry I didn't explain it clearly. I am doing what you mentioned in your suggestion. 

I have a timer that when the time is up would goto callback and there it would change the state of a global variable named `state`. in my main task inside an infinite loop, I have made a switch case that would be controlled based on this variable `state`. normally it's in idle state.
once the timer is up the state changes to store the values in a structure, after that is done I change the state which results in performing minor calculations, then to the next state which is to append the values. finally after the append is done I change the state back to idle waiting for the timer to change it again.

This main loop repeats every 10ms using `vTaskDelay(pdMS_TO_TICKS(10));` mostly in idle state where I just do a `break;`

**Reply by mobizt** [reply]
You should not use delay or vTaskDelay in the main loop.

The delay I mentioned is the delay required for an infinite loop inside the FreeRTOS task only.


No matter what types of ISRs and callback functions, you have follow my recommendation which all embedded device programmer should comply and understand why it should be avoided because of stack overflow.

**Reply by mobizt** [reply]
> in my main task inside an infinite loop, 

This is not clear what you call the main task and infinite loop.

You should create the simple code that replicates the issue to make it clear.

**Reply by mobizt** [reply]
The main task for me is only setup() or loop().

**Reply by roshanyadav29** [reply]
Again I am sorry, wrong choice of words. You are correct, main task is setup() and loop(). I declare all the functions having tasks inside the setup() and my loop() is empty. Now one of the tasks inside setup (again sorry for the previous bad choice of words) has a `for(;;)` which has a switch case statement.

```
void thread_process(void* arg) {

  for (;;) {
    switch (state) {
      case process_IDLE:
        break;
      case process_READ_Sensor:

        Serial.printf(FILE_TAG "Read Sensor data, state -> %d\n", state);

.
.
.
.
        process_val.aht_temp.temp = getData_temperature();
        process_val.aht_temp.rh = getData_humidity();

        state = process_AQI_Sensor;  // goto publish sensor data
        break;

      case process_AQI_Sensor:
#if __USE_SERIAL_DEBUG__
        Serial.printf(FILE_TAG "Calculate AQI from Sensor data, state -> %d\n", state);
#endif

// ** some functions **

        state = process_POST_Sensor;
        break;

      case process_POST_Sensor:
#if __USE_SERIAL_DEBUG__
        Serial.printf(FILE_TAG "Publish Sensor data, state -> %d\n", state);
#endif
        if (check_sheets_ready()) {
          append_values_aht(__date__, __time__, process_val.aht_temp.temp, process_val.aht_temp.rh);
        }
        state = process_IDLE;
    }
    vTaskDelay(pdMS_TO_TICKS(10));  //task delay in ms
  }
}
```

Maybe this snippet might help in visualizing the approach. Again the switch is made using a `state` variable, that is controlled by a timer. So you are definitely correct on that No Networking Functions in the ISR or Callback. Sorry for the vague explanation earlier.

**Reply by roshanyadav29** [reply]
This is the timer callback that I was talking about. I forgot to add it in the above text.

```
/* Define a timer callback function */
void vProcessTimerCallback(TimerHandle_t xTimer) {
  // read everytime the timer is called back (in seconds)
  state = process_READ_Sensor;
}
```

**Reply by roshanyadav29** [reply, marked as answer]
I don't know what the problem is but,

  ```
    case process_POST_Sensor:
#if __USE_SERIAL_DEBUG__
        Serial.printf(FILE_TAG "Publish Sensor data, state -> %d\n", state);
#endif
        if (check_sheets_ready()) {
          append_values_aht(__date__, __time__, process_val.aht_temp.temp, process_val.aht_temp.rh);
        }
        state = process_IDLE;
    }
    vTaskDelay(pdMS_TO_TICKS(10));  //task delay in ms
  }
```

Instead of this approach, I created a separate task within case `process_POST_sensor` which would append the value of the sensor and then called `vTaskDelete(NULL)`. Somehow this approach works and I can add all the values to the sheets. Not sure why the previous one wasn't working
